bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthStoreHandlerImpl.java

```diff
@@ -0,0 +1,507 @@
+                    String gsonStr = gson.toJson(lastUsed);
+                    storage.put(LAST_USED.getKey(handle), gsonStr);
+                }
+            } finally {
+                storageLock.unlock();
+            }
+        }
+
+        public void put(String handle, @Nullable AccessTokenResponse accessTokenResponse) {
+            storageLock.lock();
+            try {
+                if (storage == null) {
+                    throw new IllegalStateException(EXCEPTION_STORAGE);
+                }
+
+                if (accessTokenResponse == null) {
+                    storage.put(ACCESS_TOKEN_RESPONSE.getKey(handle), (String) null);
+                } else {
+                    String gsonAccessTokenStr = gson.toJson(accessTokenResponse);
+                    storage.put(ACCESS_TOKEN_RESPONSE.getKey(handle), gsonAccessTokenStr);
+                    String gsonDateStr = gson.toJson(LocalDateTime.now());
+                    storage.put(LAST_USED.getKey(handle), gsonDateStr);
+
+                    if (!allHandles.contains(handle)) {
+                        // update all handles index
+                        allHandles.add(handle);
+                        storage.put(STORE_KEY_INDEX_OF_HANDLES, gson.toJson(allHandles));
+                    }
+                }
+            } finally {
+                storageLock.unlock();
+            }
+        }
+
+        public void put(String handle, @Nullable PersistedParams persistedParams) {
+            storageLock.lock();
+            try {
+                if (storage == null) {
+                    throw new IllegalStateException(EXCEPTION_STORAGE);
+                }
+
+                if (persistedParams == null) {
+                    storage.put(SERVICE_CONFIGURATION.getKey(handle), (String) null);
+                } else {
+                    String gsonPersistedParamsStr = gson.toJson(persistedParams);
+                    storage.put(SERVICE_CONFIGURATION.getKey(handle), gsonPersistedParamsStr);
+                    String gsonDateStr = gson.toJson(LocalDateTime.now());
+                    storage.put(LAST_USED.getKey(handle), gsonDateStr);
+                    if (!allHandles.contains(handle)) {
+                        // update all handles index
+                        allHandles.add(handle);
+                        storage.put(STORE_KEY_INDEX_OF_HANDLES, gson.toJson(allHandles));
+                    }
+                }
+            } finally {
+                storageLock.unlock();
+            }
+        }
+
+        public void removeByHandle(String handle) {
+            logger.info("Removing handle {} from storage", handle);
+            storageLock.lock();
+            try {
+                if (allHandles.remove(handle)) { // entry exists and successfully removed
+                    if (storage == null) {
+                        // just log the error and let it be, do not throw exception
+                        logger.warn(EXCEPTION_STORAGE);
```

Comment:
This method is called for every handle - you wouldn't want to have so many similar WARN messages in the log file; the are not helpful. You hence should decide whether you consider this situation an illegal state (then better throw an exception to the caller) or a valid situation (then there is no need for any log message).

Reply:
Well spotted.  throwing IllegalStateException is consistent with the logic.